rest-api/offline/sync_consent_files.py

```diff
@@ -0,0 +1,151 @@
+import csv
+import os
+import StringIO
+
+import sqlalchemy
+
+from google.appengine.ext import deferred
+import cloudstorage
+import requests
+
+from dao import database_factory
+
+
+SOURCE_BUCKET = 'ptc-uploads-all-of-us-rdr-prod'
+
+COLUMN_BUCKET_NAME = 'Bucket Name'
+COLUMN_AGGREGATING_ORG_ID = 'Aggregating Org ID'
+COLUMN_ORG_ID = 'Org ID'
+COLUMN_ORG_STATUS = 'Org Status'
+
+ORG_STATUS_ACTIVE = 'Active'
+DEFAULT_GOOGLE_GROUP = 'no_site_pairing'
+
+
+OrgData = collections.namedtuple('OrgData', (
+  'org_id',
+  'aggregate_id',
+  'bucket_name',
+))
+
+
+ParticipantData = collections.namedtuple('ParticipantData', (
+  'participant_id',
+  'google_group',
+  'org_id',
+))
+
+
+def do_sync_consent_files():
+  """
+  entrypoint
+  """
+  # TODO: pull sheet ID from config instead of environment
+  sheet_id = os.environ.get('HPO_GOOGLE_SHEET_ID')
+  if sheet_id is None:
+    raise ValueError("Missing HPO_GOOGLE_SHEET_ID")
+  org_data_map = _load_org_data_map(sheet_id)
+  for participant_data in _iter_participants_data():
+    kwargs = {
+      "source_bucket": SOURCE_BUCKET,
+      "destination_bucket": org_data_map[participant_data.org_id].bucket_name,
+      "participant_id": participant_data.participant_id,
+      "google_group": participant_data.google_group or DEFAULT_GOOGLE_GROUP
+    }
+    deferred.defer(
+      cloudstorage_copy_objects,
+      '/{source_bucket}/Participant/P{participant_id}/'.format(**kwargs),
+      '/{destination_bucket}/Participant/{google_group}/P{participant_id}/'.format(**kwargs)
+    )
+
+
+class GoogleSheetCSVReader(csv.DictReader):
+
+  def __init__(self, sheet_id, gid=0, *args, **kwds):
+    self._sheet_id = sheet_id
+    self._gid = gid
+    response = requests.get(self._get_sheet_url(self._sheet_id, self._gid))
+    csv.DictReader.__init__(self, StringIO.StringIO(response.text), *args, **kwds)
+
+  @staticmethod
+  def _get_sheet_url(sheet_id, gid):
+    return (
+      "https://docs.google.com/spreadsheets/d/{id}/export"
+      "?format=csv"
+      "&id={id}"
+      "&gid={gid}"
+    ).format(
+      id=sheet_id,
+      gid=gid
+    )
+
+
+def _load_org_data_map(sheet_id):
+  # parse initial mapping
+  org_data_map = {
+    row.get(COLUMN_ORG_ID): OrgData(
+      row.get(COLUMN_ORG_ID),
+      row.get(COLUMN_AGGREGATING_ORG_ID),
+      row.get(COLUMN_BUCKET_NAME)
+    )
+    for row in GoogleSheetCSVReader(sheet_id)
+    if row.get(COLUMN_ORG_STATUS) == ORG_STATUS_ACTIVE
+  }
+  # apply transformations that require full map
+  return {
+    org_data.org_id: _org_data_with_bucket_inheritance_from_org_data(org_data_map, org_data)
+    for org_data in org_data_map.values()
+  }
+
+
+def _org_data_with_bucket_inheritance_from_org_data(org_data_map, org_data):
+  return OrgData(
+    org_data.org_id,
+    org_data.aggregate_id,
+    (
+      org_data.bucket_name
+      if org_data.bucket_name
+      else org_data_map[org_data.aggregate_id].bucket_name
+    )
+  )
+
+
+PARTICIPANT_DATA_SQL = sqlalchemy.text("""
```

Comment:
We can add a check for is_ghost_id from participant table (it’s brand new). And can also check for the test email pattern (@example.com) sorry I didn’t have that to begin with.

Reply:
This is applied in a newer commit. The `not like` `WHERE` clause will likely have an impact on this participant query speed.